src/Battlescape/AIModule.h

```diff
@@ -139,6 +141,29 @@ class AIModule
 	BattleUnit* getTarget();
 	/// Frees up the destination node for another Unit to select
 	void freePatrolTarget();
+	/// Checks whether anyone on our team can see the target
+	bool visibleToAnyFriend(BattleUnit *target, bool ignoreMyself = false);
+	/// Handles behavior of brutalAI
+	void brutalThink(BattleAction *action);
```

Comment:
What is your recommendation then? My understanding was to keep the code separate, which to me means that there indeed will be duplicates that do almost the same in just a slightly different way.
I don't know how to avoid duplicate code and at the same time not change the default-behaviour.

Reply:
Its hard to say, especially in in current state where when only thing I see is 1000 line blob of new code that have unspecific amount of old code included. Most time I remove code like this, and have more general version.

Most important is that that all code that do heavy lifting is not duplicated and instead of multiple magical `if` and `switch` make parameters.

Something like I handled damage types, you can compare OXC with OXCE, in reality in OXCE there is only one damage type but it fully parametrizable and because of this can recreate any damage type form OXC.

This mean for you case, I would remove your `brutalAi` and same way `sneakiAi` as both are IMHO wrong approaches and correct is expand `ai:` configure for more parameters that control alien AI.
Like you try overwrite `_aiUseDelay ` where this option 100% fully thing you need.

To sum up. You should not make one "brutalAi", you should make multiple new independed options to AI that can be enabled at once to make user MOD named "brutalAI".

Reply:
And another important thing, main concern of OXCE is modding community and making bigger and better mods.
This mean how this new AI help this? In no way, as this only make mods harder to balance as depending on choose AI alien will behave different and outcome of battle will be different.

Correct solution is allow for modder to choose exactly what and when in AI will be done, as in one battle some aliens could use new AI or old. This will expand mod Depth a looot.

Not mention interaction with difficulty levels.

Reply:
I see it more from the perspective of a player than from the perspective of a modder.
Basically more like how I'd use a difficulty-setting.
As a player I don't want to install a mod in order to change difficulty-settings. I just want to choose it and have it change the difficulty of whatever mod I want to play.

In Civ 4 or Remnants of the Precursors it also works like that. Several AIs to play against and you pick which one when you start a game.

Also as a Modder, my wish would be that the AI supports the things that I mod into the game and knows how to use them without me having to do too much about it. I wouldn't want to set 20 different variable to finetune the AI.

And from the AI-programmer's perspective, it would also be really hard to keep things modular. Especially since several things in my AI depend on each other.
In the base AI it is structured in a task-based-way. Imho that flow is complicated and hard to follow.
It fills variables for several kinds of tasks and then does a chance-based selection on which of these tasks get executed.
The methods communicate with each other by filling out member-variables and depend on the right order or sometimes are mutually exculsive as they overwrite what the other method has set.
But some are helper methods that do decision making. I think I can identify that some of these methods have different authors than others.

I would have designed the AI-code very differently from ground up. However, it fulfills the purpose of emulating the behavior of the original game well enough. And the easiest way to preserve that is to leave that code untouched.

Some of it's methods almost do what I'd like them to do. Just not quite. And I thought it would simply be cleaner to not mess them up with new flips and switches, especially not a different one for every little detail, and instead reimplement them and make my changes there.

I liked that Ray Fowler (author of Remnants of the precursors) basically said: "Here's a new folder for your AI. As long as you keep your changes separate from mine, you can do with it, whatever you want as it's optional anyways."

I'm fine with justified criticism about stuff like abusing methods in a way they aren't supposed to be used.
But when I'm asked to adjust my entire approach-of AI-development to make every single behavioral deviation from the default modable as otherwise it wouldn't even be accepted for a feature-branch that's all about AI, I guess this shifts the entire thing too much into the direction of "work" and away from "fun".

Reply:
I do not request "modularity" only "orthogonality", this mean every independent change in AI should have impended parameter, in older version of your code at least half of it could be separated not mention ones that already redundant as there were already parameters controlling it.

For end user, OXC already provide some some default mods, this could be another.

Another thing if you manage make it orthogonal then you could even split this whole feature in smaller parts, and could be possible that instead 0% or 100% but 50% of your work will be included in OXCE.

Most thing I would require in your PR do not require lot of work, mostly boilerplate to propagate parameters form rule files to ai code. And instead of `Option::brutalAi` will be `mod->getAi().featureA`, you should be able to determine what features your code provide. This will be too great documentation tools, if in two places same parameters is used then it will suggest both places are strongly linked with each other.

To conclude my hard requirements are:
* do not use global `Option::brutalAi`
* read configs from `ai:` yaml node (similar to `_aiUseDelayBlaster`).
* use one plain `struct AiType;` with members for all orthogonal features than contribute to your current  `Option::brutalAi`

This approach make it more future proof. Additionally when Ai module make cache of it of like `AiType _aiType;` it make even more opportunities to expose this confgis to scripts and allow dynamically alter some settings during battle.
Another opportunity would be in future move all `aiUseDelay*` from `Mod` to this new struct and simplify handling of some current features.